### PR TITLE
TandemFoil Paper: exact TF champion config (Lion+gc=0.5+EMA) — first test on TFP

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The TandemFoil Paper (TFP) benchmark is fresh territory — PR #2979 established the first baseline at **val_primary/field_mse = 0.00434** using AdamW + no-EMA + T_max=150 + 3L/192d. However, the **TandemFoil champion config** (Lion lr=1.25e-4, gc=0.5, EMA=0.999, T_max=10) has never been tested on TFP. Jin's PR #2947 used Lion but with different hyperparameters (gc=1.0, no-EMA) — not the exact winner.

On the standard TandemFoil benchmark, this combination (Lion + low lr + tight grad-clip + EMA) achieved **val_primary/surface_pressure_mae = 22.537** (PR #2924) — the best on that dataset. The hypothesis is that the same combination will transfer to TFP and outperform the AdamW baseline, since Lion's slower, momentum-driven convergence combined with EMA smoothing tends to find better minima on complex aerodynamic fields.

## Experiment

Test the **exact TF champion hyperparameter config** on the TandemFoil Paper dataset with 3L/192d architecture (same as the TFP champion):

| Hyperparameter | Value | Rationale |
|---|---|---|
| Optimizer | Lion | TF champion optimizer |
| lr | 1.25e-4 | TF champion lr |
| cosine-t-max | 10 | TF champion schedule |
| grad-clip | 0.5 | TF champion gc (vs gc=1.0 in PR #2947) |
| weight-decay | 1e-2 | Standard with Lion |
| EMA decay | 0.999 | TF champion EMA (not used in PR #2979 baseline) |
| model-layers | 3 | TFP champion depth |
| model-hidden-dim | 192 | TFP champion width |
| model-heads | 3 | TFP champion heads |
| enable-fourier | yes | TFP champion flag |
| All physics flags | yes | See command below |

## Current Baselines (targets to beat)

| Dataset | Metric | Current Best | PR |
|---|---|---|---|
| TandemFoil Paper | val_primary/field_mse | **0.00434** | #2979 (haku, AdamW ep199) |
| TandemFoil | val_primary/surface_pressure_mae | 22.537 | #2924 (Lion champion) |

**Primary target: beat 0.00434 on TandemFoil Paper.**

## Instructions

Run the following command. Use `--wandb_group tandemfoil-paper-lion-sweep` so future Lion variants on TFP group together in W&B:

```bash
cd target/icml2026
python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --ema-decay 0.999 \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --wandb_group tandemfoil-paper-lion-sweep \
  --epochs 999
```

## What to report

Report `val_primary/field_mse` at **best validation epoch** (not terminal epoch). Also report:
- `val_primary/surface_pressure_mae` if available
- The epoch at which the best checkpoint was found
- W&B run ID

## Key data pipeline notes (from PR #2979)

Three bugs were fixed in the TFP data pipeline in PR #2979 — these fixes are already in the `radford` branch you branched from:
1. AoA validation too strict in `split_paper_experiment4.py:191` — fixed
2. Non-finite pressure values clamped via `torch.where` in `core/datasets.py:142` — fixed
3. Stats computation masked non-finite values in `split_paper_experiment4.py:262` — fixed

These fixes are in the codebase — no action needed, just confirming they are present.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT haku:

## Results

### Best Validation Result (epoch 443)

| Metric | Value |
|--------|-------|
| **val_primary/field_mse** | **0.00238** |
| val/surface_mse | 0.001517 |
| val/surface_mse_Ux | 0.001170 |
| val/surface_mse_Uy | 0.003333 |
| val/surface_mse_p | 4.81e-05 |
| val/volume_mse | 0.002397 |

### Comparison to Baseline

| Metric | Baseline (PR #2979) | This Run | Change |
|--------|---------------------|----------|--------|
| val_primary/field_mse | **0.00434** | **0.00238** | **-45.1%** |

### Run Details

- **W&B run:** d1xh0o1p (`haku/tfp-lion-champion-config`)
- **Best epoch:** 443 (of 469 before divergence)
- **Peak GPU memory:** ~40.7 GB (GPU 0, RTX PRO 6000)
- **Total epochs:** 469 stable + divergence at ep462 (grad norms: 5.8 → 120 → 1656 → 4444)
- **Training budget:** 360 min (SENPAI_TIMEOUT_MINUTES), 999 max epochs

### Command

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --ema-decay 0.999 \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --wandb-group tandemfoil-paper-lion-sweep \
  --agent haku \
  --wandb-name "haku/tfp-lion-champion-config" \
  --epochs 999
```

### What Happened

**The Lion champion config decisively beats AdamW on TandemFoil Paper.** The hypothesis is confirmed — the exact TF champion combination (Lion lr=1.25e-4, gc=0.5, EMA=0.999, T_max=10) transfers strongly to TFP, achieving a 45.1% improvement over the AdamW baseline.

Key observations:
1. **Consistent descent over 460 epochs:** The T_max=10 cosine cycling found progressively deeper basins across ~46 full cycles. The trough envelope descended smoothly from 0.271 (ep1) → 0.036 (ep10) → 0.007 (ep90) → 0.004 (ep170) → 0.003 (ep260) → **0.00238** (ep443).
2. **EMA warmup:** Actual EMA decay ramped from 0.976 (ep1) to 0.999 (fully warm by ep83), at which point the model's descent accelerated.
3. **gc=0.5 stability:** Gradient clipping at 0.5 kept the model stable for 460+ epochs — far longer than the gc=1.0 divergence at ep167 seen on standard TandemFoil (PR #2924). All 360 batches were clipped initially; this decreased to ~315-330 by late epochs as the model settled.
4. **Late divergence at ep462:** Grad norms exploded from ~5.8 to 120 → 1656 → 4444. This is the same T_max=10 instability pattern seen across datasets — the cosine cycling eventually triggers irreversible instability. The EMA model preserves the ep443 best checkpoint.
5. **Still descending at divergence:** The model was still finding new bests (ep443 vs ep435 vs ep415) right up to the divergence event. More training with stronger gradient clipping or a warmup schedule could push even deeper.

### Suggested Follow-ups

1. **gc=0.3 or gc=0.25:** Even tighter gradient clipping might extend stable training past ep462, allowing the model to descend further.
2. **Lion lr sweep on TFP:** The TF champion used lr=1.25e-4 but lower LR (e.g., 1e-4, 8e-5) might push deeper on TFP given the different geometry complexity.
3. **T_max=20 or T_max=30 on TFP with Lion:** Longer cosine periods might avoid the late divergence while maintaining deep basin exploration.
4. **4L/256d + Lion on TFP:** The 3L/192d architecture was inherited from the TFP champion — testing Lion with a larger architecture could unlock further gains.

**Comment by tcapelle** [comment]
ADVISOR: Merging — this is a decisive 45.1% improvement on TFP (0.002383 vs 0.00434 baseline). The TF champion config (Lion lr=1.25e-4, gc=0.5, EMA=0.999, T_max=10) transfers strongly to TFP. This establishes a new TFP baseline. Excellent work, haku.
